pkg/providers/amifamily/resolver.go

```diff
@@ -134,30 +136,62 @@ func (r DefaultResolver) Resolve(nodeClass *v1.EC2NodeClass, nodeClaim *karpv1.N
 		// In order to support reserved ENIs for CNI custom networking setups,
 		// we need to pass down the max-pods calculation to the kubelet.
 		// This requires that we resolve a unique launch template per max-pods value.
-		// Similarly, instance types configured with EfAs require unique launch templates depending on the number of
+		// Similarly, instance types configured with EFAs require unique launch templates depending on the number of
 		// EFAs they support.
+		// Reservations IDs are also included since we need to create a separate LaunchTemplate per reservation ID when
+		// launching reserved capacity. If it's a reserved capacity launch, we've already filtered the instance types
+		// further up the call stack.
 		type launchTemplateParams struct {
-			efaCount int
-			maxPods  int
+			efaCount       int
+			maxPods        int
+			reservationIDs string
 		}
-		paramsToInstanceTypes := lo.GroupBy(instanceTypes, func(instanceType *cloudprovider.InstanceType) launchTemplateParams {
+		paramsToInstanceTypes := lo.GroupBy(instanceTypes, func(it *cloudprovider.InstanceType) launchTemplateParams {
 			return launchTemplateParams{
 				efaCount: lo.Ternary(
 					lo.Contains(lo.Keys(nodeClaim.Spec.Resources.Requests), v1.ResourceEFA),
-					int(lo.ToPtr(instanceType.Capacity[v1.ResourceEFA]).Value()),
+					int(lo.ToPtr(it.Capacity[v1.ResourceEFA]).Value()),
 					0,
 				),
-				maxPods: int(instanceType.Capacity.Pods().Value()),
+				maxPods: int(it.Capacity.Pods().Value()),
+				// If we're dealing with reserved instances, there's only going to be a single instance per group. This invariant
+				// is due to reservation IDs not being shared across instance types. Because of this, we don't need to worry about
+				// ordering in this string.
+				reservationIDs: lo.Ternary(
+					capacityType == karpv1.CapacityTypeReserved,
+					strings.Join(selectReservationIDs(it, nodeClaim), ","),
+					"",
+				),
 			}
 		})
+
 		for params, instanceTypes := range paramsToInstanceTypes {
-			resolved := r.resolveLaunchTemplate(nodeClass, nodeClaim, instanceTypes, capacityType, amiFamily, amiID, params.maxPods, params.efaCount, options)
-			resolvedTemplates = append(resolvedTemplates, resolved)
+			reservationIDs := strings.Split(params.reservationIDs, ",")
+			resolvedTemplates = append(resolvedTemplates, r.resolveLaunchTemplates(nodeClass, nodeClaim, instanceTypes, capacityType, amiFamily, amiID, params.maxPods, params.efaCount, reservationIDs, options)...)
 		}
 	}
 	return resolvedTemplates, nil
 }
 
+// selectReservationIDs filters the set of reservation IDs available on the given instance type to only include those
```

Comment:
Note: going to be dropping this function since we now do the same thing further up the stack when filtering reserved offerings - this was required so we could have a 1:1 mapping between reservation IDs and instance pools when identifying which reservation ran out of capacity based on a fleet error.